connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorRestartApiIntegrationTest.java

```diff
@@ -0,0 +1,425 @@
+        connect.start();
+
+        // Call the Restart API
+        String restartEndpoint = connect.endpointForResource(
+                String.format("connectors/%s/restart?onlyFailed=" + onlyFailed + "&includeTasks=" + includeTasks, connectorName));
+        Response response = connect.requestPost(restartEndpoint, "", Collections.emptyMap());
+        assertEquals(Response.Status.NOT_FOUND.getStatusCode(), response.getStatus());
+    }
+
+    @Test
+    public void testRestartOnlyConnector() throws Exception {
+        runningConnectorAndTasksRestart(false, false, 1, allTasksExpectedRestarts(0), false);
+    }
+
+    @Test
+    public void testRestartBoth() throws Exception {
+        runningConnectorAndTasksRestart(false, true, 1, allTasksExpectedRestarts(1), false);
+    }
+
+    @Test
+    public void testRestartNoopOnlyConnector() throws Exception {
+        runningConnectorAndTasksRestart(true, false, 0, allTasksExpectedRestarts(0), true);
+    }
+
+    @Test
+    public void testRestartNoopBoth() throws Exception {
+        runningConnectorAndTasksRestart(true, true, 0, allTasksExpectedRestarts(0), true);
+    }
+
+    @Test
+    public void testFailedConnectorRestartOnlyConnector() throws Exception {
+        failedConnectorRestart(false, false, 1);
+    }
+
+    @Test
+    public void testFailedConnectorRestartBoth() throws Exception {
+        failedConnectorRestart(false, true, 1);
+    }
+
+    @Test
+    public void testFailedConnectorRestartBoth2() throws Exception {
+        failedConnectorRestart(true, true, 1);
+    }
+
+    @Test
+    public void testFailedTasksRestartOnlyConnector() throws Exception {
+        failedTasksRestart(false, false, 1, allTasksExpectedRestarts(0), buildAllTasksToFail(), false);
+    }
+
+    @Test
+    public void testFailedTasksRestartOnlyTasks() throws Exception {
+        failedTasksRestart(true, true, 0, allTasksExpectedRestarts(1), buildAllTasksToFail(), false);
+    }
+
+    @Test
+    public void testFailedTasksRestartNoop() throws Exception {
+        failedTasksRestart(true, false, 0, allTasksExpectedRestarts(0), buildAllTasksToFail(), true);
+    }
+
+    @Test
+    public void testFailedTasksRestartBoth() throws Exception {
+        failedTasksRestart(false, true, 1, allTasksExpectedRestarts(1), buildAllTasksToFail(), false);
+    }
+
+    @Test
+    public void testOneFailedTasksRestartOnlyOneTasks() throws Exception {
+        Set<String> tasksToFail = Collections.singleton(taskId(1));
+        failedTasksRestart(true, true, 0, buildExpectedTasksRestarts(tasksToFail), tasksToFail, false);
+    }
+
+    @Test
+    public void testMultiWorkerRestartOnlyConnector() throws Exception {
+        runningConnectorAndTasksRestart(false, false, 1, allTasksExpectedRestarts(0), false, 4);
+    }
+
+    @Test
+    public void testMultiWorkerRestartBoth() throws Exception {
+        runningConnectorAndTasksRestart(false, true, 1, allTasksExpectedRestarts(1), false, 4);
+    }
```

Comment:
IIUC, we are not actually testing that the connector and task instances are distributed across different worker nodes, though it is likely to happen given the light connector load on the Connect cluster.

Reply:
Could you guide me if there a predictable way to spread the tasks then I can change the test?

I found that when I gave numWorkers>1 most of the time they were distributed that's how I found an NPE bug and had to use @DefaultValue in the API.

I am using numWorkers=1 in other test to have a deterministic number of task Restarts, whenever I had numWorkers>1 the tasks would get restarted more than what I expect due to rebalance on worker nodes coming up.

Reply:
Two things. First, these multi-worker ITs are not necessarily testing the requests getting forwarded to the leader, but that's really hard to do. Second, it might be good to try to send the request to a worker that is not running any instances to be restarted. One way to do that is to run 1 more worker than the total number of `Connector` and `Task` instances for the connector being restarted. Though I'm not sure this adds a lot of value given the others unit tests.

Reply:
@rhauch this is an excellent suggestion, I just added this test based on the above discussion https://github.com/apache/kafka/pull/10822/commits/936e9da9f38ce6b87c52ff6ef95558c75bf56982 and now feel more confident that we are testing the distributed behavior of the feature.

Here is a sample response I copied from the above test output and we can see the call from the test was made on a worker not running the task or connector instance.

> [2021-06-17 11:41:22,306] INFO POST response for URL=http://localhost:49544/connectors/conn-testMultiWorkerRestartBothConnectorAndTasks/restart?onlyFailed=false&includeTasks=true is {"name":"conn-testMultiWorkerRestartBothConnectorAndTasks","connector":{"state":"RESTARTING","worker_id":"localhost:49424"},"tasks":[{"id":0,"state":"RESTARTING","worker_id":"localhost:49424"},{"id":1,"state":"RESTARTING","worker_id":"localhost:49452"},{"id":2,"state":"RESTARTING","worker_id":"localhost:49484"},{"id":3,"state":"RESTARTING","worker_id":"localhost:49512"}],"type":"source"} (org.apache.kafka.connect.util.clusters.EmbeddedConnectCluster:735)